File: src/utils/numberToWords.ts
// Converts a KWD amount into the "[amount] Kuwaiti Dinars only" style
// phrase used by the lettered quotation/contract templates. Correct for
// the range these documents realistically use (up to low millions);
// not a general-purpose currency-to-words library.

const AR_ONES = [
  '',
  'واحد',
  'اثنان',
  'ثلاثة',
  'أربعة',
  'خمسة',
  'ستة',
  'سبعة',
  'ثمانية',
  'تسعة',
  'عشرة',
  'أحد عشر',
  'اثنا عشر',
  'ثلاثة عشر',
  'أربعة عشر',
  'خمسة عشر',
  'ستة عشر',
  'سبعة عشر',
  'ثمانية عشر',
  'تسعة عشر',
]
const AR_TENS = ['', '', 'عشرون', 'ثلاثون', 'أربعون', 'خمسون', 'ستون', 'سبعون', 'ثمانون', 'تسعون']
const AR_HUNDREDS = [
  '',
  'مائة',
  'مئتان',
  'ثلاثمائة',
  'أربعمائة',
  'خمسمائة',
  'ستمائة',
  'سبعمائة',
  'ثمانمائة',
  'تسعمائة',
]

function arabicUnderThousand(n: number): string {
  if (n === 0) return ''
  const parts: string[] = []
  const hundreds = Math.floor(n / 100)
  const remainder = n % 100
  if (hundreds) parts.push(AR_HUNDREDS[hundreds])
  if (remainder) {
    if (remainder < 20) {
      parts.push(AR_ONES[remainder])
    } else {
      const tens = Math.floor(remainder / 10)
      const ones = remainder % 10
      parts.push(ones ? `${AR_ONES[ones]} و${AR_TENS[tens]}` : AR_TENS[tens])
    }
  }
  return parts.join(' و')
}

/** Whole-number integer (0 to low millions) to Arabic words. */
export function integerToArabicWords(value: number): string {
  const n = Math.floor(Math.abs(value))
  if (n === 0) return 'صفر'

  const millions = Math.floor(n / 1_000_000)
  const thousands = Math.floor((n % 1_000_000) / 1000)
  const rest = n % 1000

  const groups: string[] = []
  if (millions) {
    groups.push(millions === 1 ? 'مليون' : millions === 2 ? 'مليونان' : `${arabicUnderThousand(millions)} مليون`)
  }
  if (thousands) {
    groups.push(thousands === 1 ? 'ألف' : thousands === 2 ? 'ألفان' : `${arabicUnderThousand(thousands)} ألف`)
  }
  if (rest) {
    groups.push(arabicUnderThousand(rest))
  }
  return groups.join(' و')
}

/** e.g. 1250.500 KWD -> "ألف ومئتان وخمسون ديناراً كويتياً و500 فلس فقط لا غير" */
export function amountToArabicWords(amount: number, currencyLabel = 'دينار كويتي'): string {
  const dinars = Math.floor(amount)
  const fils = Math.round((amount - dinars) * 1000)
  const dinarsWords = integerToArabicWords(dinars)
  const dinarSuffix = dinars === 1 ? 'دينار كويتي' : dinars === 2 ? 'ديناران كويتيان' : `${currencyLabel}اً`
  const base = `${dinarsWords} ${dinarSuffix}`
  const filsPart = fils > 0 ? ` و${fils} فلس` : ''
  return `(${base}${filsPart} فقط لا غير)`
}

const EN_ONES = [
  '',
  'One',
  'Two',
  'Three',
  'Four',
  'Five',
  'Six',
  'Seven',
  'Eight',
  'Nine',
  'Ten',
  'Eleven',
  'Twelve',
  'Thirteen',
  'Fourteen',
  'Fifteen',
  'Sixteen',
  'Seventeen',
  'Eighteen',
  'Nineteen',
]
const EN_TENS = ['', '', 'Twenty', 'Thirty', 'Forty', 'Fifty', 'Sixty', 'Seventy', 'Eighty', 'Ninety']

function englishUnderThousand(n: number): string {
  if (n === 0) return ''
  const parts: string[] = []
  const hundreds = Math.floor(n / 100)
  const remainder = n % 100
  if (hundreds) parts.push(`${EN_ONES[hundreds]} Hundred`)
  if (remainder) {
    if (remainder < 20) {
      parts.push(EN_ONES[remainder])
    } else {
      const tens = Math.floor(remainder / 10)
      const ones = remainder % 10
      parts.push(ones ? `${EN_TENS[tens]}-${EN_ONES[ones]}` : EN_TENS[tens])
    }
  }
  return parts.join(' ')
}

export function integerToEnglishWords(value: number): string {
  const n = Math.floor(Math.abs(value))
  if (n === 0) return 'Zero'

  const millions = Math.floor(n / 1_000_000)
  const thousands = Math.floor((n % 1_000_000) / 1000)
  const rest = n % 1000

  const groups: string[] = []
  if (millions) groups.push(`${englishUnderThousand(millions)} Million`)
  if (thousands) groups.push(`${englishUnderThousand(thousands)} Thousand`)
  if (rest) groups.push(englishUnderThousand(rest))
  return groups.join(' ')
}

/** e.g. 1250.500 KWD -> "Kuwaiti Dinars One Thousand Two Hundred and Fifty and 500 Fils Only" */
export function amountToEnglishWords(amount: number, currencyLabel = 'Kuwaiti Dinars'): string {
  const dinars = Math.floor(amount)
  const fils = Math.round((amount - dinars) * 1000)
  const dinarsWords = integerToEnglishWords(dinars)
  const filsPart = fils > 0 ? ` and ${fils} Fils` : ''
  return `(${currencyLabel} ${dinarsWords}${filsPart} Only)`
}

File: backend/app/core/number_to_words.py
"""Spells out a currency amount in English words (Decimal(1500.500) ->
"One Thousand Five Hundred Kuwaiti Dinars and Five Hundred Fils Only") --
Kuwaiti Dinar quotations traditionally state the total this way, and this
is exactly the "amount_in_words" placeholder an uploaded Quotation/
Contract document template can bind to (see
document_template_service.MERGE_FIELD_CATALOG). No third-party dependency
(num2words etc.) -- amounts here are always a plain currency total, well
within what a small hand-written converter handles correctly, and it
keeps the currency-name/minor-unit table (KWD's minor unit is "Fils", not
"Cents") fully under our control.
"""

from decimal import ROUND_HALF_UP, Decimal

_ONES = (
    "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine",
    "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen",
)
_TENS = ("", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety")
_SCALES = (
    (1_000_000_000, "Billion"),
    (1_000_000, "Million"),
    (1_000, "Thousand"),
    (100, "Hundred"),
)

# (major unit name, minor unit name, minor units per major unit) -- KWD's
# 3-decimal Fils is why this table exists at all rather than a flat "/100"
# assumption; every currency the Quotation/Contract form offers (see
# NewQuotationDialog.vue's CURRENCY_OPTIONS) is listed explicitly rather
# than guessed, so an unlisted currency fails loudly instead of silently
# mis-converting.
_CURRENCY_UNITS: dict[str, tuple[str, str, int]] = {
    "KWD": ("Kuwaiti Dinar", "Fils", 1000),
    "USD": ("US Dollar", "Cent", 100),
    "AED": ("UAE Dirham", "Fils", 100),
    "EUR": ("Euro", "Cent", 100),
}

# Minor-unit names that are already invariant between singular and plural
# in conventional English usage (Gulf currencies' "Fils" chief among
# them) -- pluralizing these by appending "s" would read as a typo
# ("Filss"), not a plural.
_INVARIANT_MINOR_UNITS = {"Fils"}


def _int_to_words(value: int) -> str:
    if value == 0:
        return "Zero"
    if value < 20:
        return _ONES[value]
    if value < 100:
        tens, remainder = divmod(value, 10)
        return _TENS[tens] + (f" {_ONES[remainder]}" if remainder else "")
    for scale_value, scale_name in _SCALES:
        if value >= scale_value:
            quotient, remainder = divmod(value, scale_value)
            head = f"{_int_to_words(quotient)} {scale_name}"
            return f"{head} {_int_to_words(remainder)}" if remainder else head
    return _ONES[value]  # pragma: no cover -- unreachable, every branch above covers 0-999,999,999


def amount_to_words(amount: Decimal, currency: str) -> str:
    """"Only" is appended the way Kuwaiti/Gulf commercial documents
    conventionally close this line, signalling the figure isn't open to a
    trailing-digits rounding dispute. Rounds to the currency's smallest
    unit first (banker's-rounding-free, matching how the amount itself is
    already stored/displayed -- see quotation_service.compute_amount)
    rather than truncating, so e.g. 1500.999 KWD reads as "...and 999
    Fils", not 998 from float error."""
    if currency not in _CURRENCY_UNITS:
        raise ValueError(f"amount_to_words does not know currency {currency!r}")
    major_name, minor_name, minor_per_major = _CURRENCY_UNITS[currency]

    scaled = int((amount * minor_per_major).quantize(Decimal(1), rounding=ROUND_HALF_UP))
    major_units, minor_units = divmod(scaled, minor_per_major)

    major_plural = major_name if major_units == 1 else f"{major_name}s"
    words = f"{_int_to_words(major_units)} {major_plural}"
    if minor_units:
        minor_plural = minor_name if minor_units == 1 or minor_name in _INVARIANT_MINOR_UNITS else f"{minor_name}s"
        words += f" and {_int_to_words(minor_units)} {minor_plural}"
    return f"{words} Only"
